### PR TITLE
Implmement a project system for Razor

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        {
+            for (var i = 0; i < _listeners.Count; i++)
+            {
+                if (_listeners[i].TryGetTarget(out var listener))
+                {
+                    listener.Notify(e);
+                }
+                else
+                {
+                    _listeners.RemoveAt(i--);
```
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: Love me an `i--` 😉

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManagerFactory.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    [Shared]
```
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManagerFactory.cs]: What's this guy do?

File: src/Microsoft.VisualStudio.LanguageServices.Razor/Legacy/LegacyTemplateEngineFactoryService.cs [hunk 3]
```diff
@@ -3,8 +3,11 @@
```
Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/Legacy/LegacyTemplateEngineFactoryService.cs]: Make sure not to commit this file in the initial push.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManagerFactory.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManagerFactory.cs]: Marks it as a singleton. We should probably have this on more things.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        private void InitializeSolution(Solution solution)
+        {
+            foreach (var kvp in _projects.ToArray())
+            {
+                _projects.Remove(kvp.Key);
+                NofifyListeners(new ProjectChangeEventArgs(kvp.Value, ProjectChangeKind.Removed));
+            }
+
+            if (solution != null)
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: I was being defensive here, but this won't ever be null. I'll change to an assert.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: RIP

Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: I thought about just going backwards, but I thought you'd enjoy it if I memorialized "the taylor"

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                        {
+                            _projects.Remove(e.ProjectId);
+                            NofifyListeners(new ProjectChangeEventArgs(snapshot, ProjectChangeKind.Removed));
+                        }
+
+                        break;
+                    }
+
+                case WorkspaceChangeKind.SolutionAdded:
+                case WorkspaceChangeKind.SolutionChanged:
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: Two per project really in the case of a reload. Generally a subscriber will filter by path.

I'm not super concerned about any of these besides SolutionChanged. VS will close all of the editors if you do these other things that reload the solution.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                case WorkspaceChangeKind.SolutionCleared:
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: Cleared means you still have a solution, but it's been emptied of projects. Removed means you don't have a solution.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: About SolutionChanged..... MSDN says... `

The current solution changed for an unspecified reason.
--

`

Thanks MSDN. Great

[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: Solution Crawler treats it the same as a reload and I think we should too https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.cs#L326

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                case WorkspaceChangeKind.SolutionReloaded:
+                case WorkspaceChangeKind.SolutionRemoved:
+                    InitializeSolution(e.NewSolution);
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: In VS, yeah I think so

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManagerFactory.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManagerFactory.cs]: Makes it a singleton

File: src/Microsoft.VisualStudio.LanguageServices.Razor/Legacy/LegacyTemplateEngineFactoryService.cs (hunk 3 above)
Comment (rynowak) [reply src/Microsoft.VisualStudio.LanguageServices.Razor/Legacy/LegacyTemplateEngineFactoryService.cs]: Yeah will undo. Here lies my attempt to do this in a non-stateful way.

**Comment by rynowak** [comment]
Updated

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (NTaylorMullen) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: Alright, wanted to make sure adding a file to a project didn't result in a hierarchy of events getting triggered and then firing this one.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (NTaylorMullen) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: In the Cleared case are there ever any projects?

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            foreach (var project in solution.Projects)
+            {
+                var projectState = new DefaultProjectSnapshot(project);
+                _projects[project.Id] = projectState;
+
+                NofifyListeners(new ProjectChangeEventArgs(projectState, ProjectChangeKind.Added));
+            }
+        }
+
+        private void NofifyListeners(ProjectChangeEventArgs e)
```
Comment (ajaybhargavb) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: `Nofify` => `Notify`

**Pull request merged**